src/main/scala/org/graphframes/lib/AggregateMessages.scala

```diff
@@ -100,22 +101,50 @@ class AggregateMessages private[graphframes] (private val g: GraphFrame)
    *   aggResult.join(g.vertices, ID)
    * }}}
    */
-  def agg(aggCol: Column): DataFrame = {
+  def agg(aggCol: Column, aggCols: Column*): DataFrame = {
+    def containsStructNameForSpark2(columnName: String, structName: String) =
+      columnName.startsWith(s"${structName}.")
+    def containsStructNameForSpark16(columnName: String, structName: String) =
+      columnName.startsWith(s"${structName}[")
+    def removeStructNameForSpark2(columnName: String, structName: String) =
+      columnName.substring(s"${structName}.".length)
+    def removeStructNameForSpark16(columnName: String, structName: String) =
+      columnName.substring(s"${structName}[".length, columnName.length - 1)
+
+    def removeStructName(columnName: String) = columnName match {
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.SRC}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.SRC}")
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.DST}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.DST}")
+      case cn if containsStructNameForSpark16(cn, s"${GraphFrame.EDGE}") =>
+        removeStructNameForSpark16(cn, s"${GraphFrame.EDGE}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.SRC}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.SRC}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.DST}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.DST}")
+      case cn if containsStructNameForSpark2(cn, s"${GraphFrame.EDGE}") =>
+        removeStructNameForSpark2(cn, s"${GraphFrame.EDGE}")
+      case cn  => cn
+    }
+    def msgColumn(df: DataFrame) = df.columns.filter(_ != ID) match {
+      case Array(c) => df.withColumnRenamed(c, AggregateMessages.MSG_COL_NAME)
+      case columns => df.select(
+        df(ID),
+        struct(columns.sorted.map(c => col(s"`${c}`").as(removeStructName(c))) :_*)
+          .as(AggregateMessages.MSG_COL_NAME))
```

Comment:
can you explain and comment why this and associated helper functions are necessary?